### ISSUE TITLE
DNN huge performance dropout on 32 bit

### ISSUE THREAD

**Issue by juacamgo** [opencv/opencv#19691]
##### System information (version)

- OpenCV => 4.5.1
- Operating System / Platform => Windows 10 Pro
- Compiler => Visual Studio 2019
- Generator => CMake 3.19.3

##### Detailed description

I am using 32 bit OpenCV 4.5.1 compiled with CMake on a project for DNN. I need to real-time process frames and use the net to check the position of a barcode on the image.

While using the 32 bit compiled version on my machine, I am getting a inference of 9-11 fps, while using the 64 bit compiled version I am getting about 29-32 fps (Using Python I get about 22-23 fps).

I wonder if maybe there is a bug or something that is causing the huge performance dropout between 32 and 64 bits.

Also I've noticed that the first frame I process it's always slower that the others. For example, using 64 bits, the first frame could be 22 fps while the others will be between 27-30 fps.

![infernece](https://user-images.githubusercontent.com/47568991/110339615-a4640e80-8028-11eb-8717-76d2aacf3007.PNG)

##### Steps to reproduce

Below is the C++ code I am using, but I think you can reproduce the problem with any net using 32 and 64 bit versions of OpenCV:

```.cpp
#include <stdbool.h>

#include <opencv2/dnn.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>

#include "barcode.h"
#include "draw.h"

using namespace cv;
using namespace dnn;

const String weights = "barcode-yolov4-tiny_final.weights";
const String cfg = "barcode-yolov4-tiny-detector.cfg";
const Mat image = imread("azteccode_test.png");

const int ITERATIONS = 25;

constexpr float CONFIDENCE_THRESHOLD = 0.2;
constexpr float NMS_THRESHOLD = 0.4;
constexpr int NUM_CLASSES = 1;

std::vector<std::string> classes;

int main()
{
	classes.push_back("barcode");

	Net net = readNetFromDarknet(cfg, weights);
	net.setPreferableBackend(DNN_BACKEND_OPENCV);
	net.setPreferableTarget(DNN_TARGET_CPU);


	DetectionModel model(net);
	model.setInputParams(1.0 / 255, Size(416, 416));

	for (int j = 0; j < ITERATIONS; j++)
	{
		Mat frame = image.clone();
		auto totalStart = std::chrono::steady_clock::now();
		MatShape classIds;
		std::vector<float> confidences;
		std::vector<Rect> boxes;

		auto dnnStart = std::chrono::steady_clock::now();
		model.detect(image, classIds, confidences, boxes, CONFIDENCE_THRESHOLD, NMS_THRESHOLD);
		auto dnnEnd = std::chrono::steady_clock::now();

		for (int i = 0; i < boxes.size(); i++)
		{
			Rect box = boxes[i];
			Scalar color = (255, 0, 0);
			String label(" % s : % f", (classes[classIds[0]], confidences[i]));
			rectangle(frame, box, color, 2);
			putText(frame, label, Point(box.x, box.y - 10), FONT_HERSHEY_COMPLEX, 0.5, color, 2);
		}

		auto totalEnd = std::chrono::steady_clock::now();

		float inferenceFps = 1000.0 / std::chrono::duration_cast<std::chrono::milliseconds>(dnnEnd - dnnStart).count();
		float totalFps = 1000.0 / std::chrono::duration_cast<std::chrono::milliseconds>(totalEnd - totalStart).count();

		printf("InferenceFps: %f fps \n", inferenceFps);
		
		std::ostringstream stats_ss;
		stats_ss << std::fixed << std::setprecision(2);
		stats_ss << "Inference FPS: " << inferenceFps << ", Total FPS: " << totalFps;
		auto stats = stats_ss.str();

		int baseline;
		auto stats_bg_sz = getTextSize(stats.c_str(), FONT_HERSHEY_COMPLEX_SMALL, 1, 1, &baseline);
		rectangle(image, Point(0, 0), Point(stats_bg_sz.width, stats_bg_sz.height + 10), Scalar(0, 0, 0), FILLED);
		putText(image, stats.c_str(), Point(0, stats_bg_sz.height + 5), FONT_HERSHEY_COMPLEX_SMALL, 1, Scalar(255, 255, 255));

		imshow("Detected", image);
		waitKey(0);
		destroyAllWindows();
	}

    return 0;
}
```

**Comment by alalek** [comment]
1. You should compare build options first. Use `cv::getBuildInformation()` for that.
2. Compare on older CPUs. Changes should be minimal for SSE4.2 instruction set.

Several optimizations are just disabled for 32-bit code, like AVX2 instructions in dispatched code, IPP optimizations.

**Comment by juacamgo** [comment]
Hello, I've compared the build information as you mentioned, and the only difference is that:

![diff1](https://user-images.githubusercontent.com/47568991/110604167-0dae6380-8188-11eb-877a-b93f999d692b.PNG)

You mentioned that several optimizations are disabled for 32-bit code. Is there a way to compile OpenCV 32 bit with some optimizations?

Thank you.

**Comment by alalek** [comment]
Adjust `CPU_BASELINE` / `CPU_DISPATCH` CMake options:
https://github.com/opencv/opencv/wiki/CPU-optimizations-build-options

**Comment by juacamgo** [comment]
Hello @alalek , first of all I want to thank you for the support!

I've adjusted my CMake options to compile using the optimizations and now build information is:
32 bit at left, 64 bit at right

![diff1](https://user-images.githubusercontent.com/47568991/110617250-be6f2f80-8195-11eb-9af9-8f35c21632aa.PNG)

I got an error compiling and finally the AVX512_SKX is not added, but even so I can use the compiled.

The point is that with the new library, I got exactly the same inference speed. Could be possible that due to my Windows version (64 bits), using the 32 bit version virtualizes and that is causing the performance issues?

I can get over 60 fps using the 64 bit compiled version and about 30 fps using the 32 bit version.
